### PR TITLE
Clear note witness caches on reindexing

### LINKED ISSUE #1394: Drop the note witness cache when reindexing
If the node is launched with `-reindex` then the wallet will get called with every single transaction we possess since genesis. We should not copy across the witness caches for those transactions in `CWallet::AddToWallet()` when this occurs, to ensure there is no cache corruption.

### PR DESCRIPTION
**Pull request by str4d**
This PR also fixes a test that was passing arguments in the wrong order.

Closes #1394

### CODE REVIEW

**Comment by str4d** [comment]
Still needs review.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -592,6 +592,17 @@ void CWallet::AddToSpends(const uint256& wtxid)
     }
 }
 
+void CWallet::ClearNoteWitnessCache()
+{
+    LOCK(cs_wallet);
+    for (std::pair<const uint256, CWalletTx>& wtxItem : mapWallet) {
+        for (mapNoteData_t::value_type& item : wtxItem.second.mapNoteData) {
+            CNoteData* nd = &(item.second);
+            nd->witnesses.clear();
```
Comment (daira) [inline src/wallet/wallet.cpp]: Why not just ``item.second.witnesses.clear()``? Or use a ``CNoteData&``?

Comment (daira) [inline src/wallet/wallet.cpp]: ^ Does not block.

**Comment by daira** [comment]
utACK. Tests look okay, have not checked coverage.

**Comment by bitcartel** [comment]
ACK. Did not check coverage.

**Comment by str4d** [comment]
I have 40 lines of test for 10 lines of code, so I'm just going to say coverage is fine ;)

@zkbot r+

**Pull request merged**